prometheus: update to 3.1.0.

bryanasdev000 opened the pull request:
<!-- Uncomment relevant sections and delete options which are not applicable -->

#### Testing the changes
- I tested the changes in this PR: **Yes, running on my homelab**

<!--
#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**|**NO**
-->

<!-- Note: If the build is likely to take more than 2 hours, please add ci skip tag as described in
https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration
and test at least one native build and, if supported, at least one cross build.
Ignore this section if this PR is not skipping CI.
-->

#### Local build testing
- I built this PR locally for my native architecture, (x86_64-glibc)

#### Note

Although 3.x is available, Prometheus 2.55 or newer "required" for a major (3.x) upgrade, per migration guide (https://prometheus.io/docs/prometheus/3.0/migration/#tsdb-format-and-downgrade).

bryanasdev000 commented:
Also, a question, for the 3.x release, a separated packaged would be needed? 

Current release is 2.55.1, current LTS is 2.53.3, and now 3.0.0.

bryanasdev000 commented:
Is possible to retry tests on i686?

Did not find an error.

TinfoilSubmarine commented:
> Although 3.x is available, Prometheus 2.55 or newer "required" for a major (3.x) upgrade, per migration guide (https://prometheus.io/docs/prometheus/3.0/migration/#tsdb-format-and-downgrade).

Are you sure this assumption is correct? The linked guide doesn't say anything about needing >=2.55 to upgrade to v3. It only says that >=2.55 is needed when downgrading from v3.

bryanasdev000 commented:
> > Although 3.x is available, Prometheus 2.55 or newer "required" for a major (3.x) upgrade, per migration guide (https://prometheus.io/docs/prometheus/3.0/migration/#tsdb-format-and-downgrade).
> 
> Are you sure this assumption is correct? The linked guide doesn't say anything about needing >=2.55 to upgrade to v3. It only says that >=2.55 is needed when downgrading from v3.

Bad wording on my part, sorry.

AFAIK is not required to have 2.55 to upgrade (thought the two upgrades that I did, I was on 2.55), in this case I put it as "required" and the asked how to proceed in void packages because is a major upgrade, and as I have understood, there will be multiple versions being supported at the same time (at least 2, if I understood correctly, the 2.x LTS and 3.x now).

So in case things go wrong, users have a way to downgrade and recover.

TinfoilSubmarine commented:
>  as I have understood, there will be multiple versions being supported at the same time (at least 2, if I understood correctly, the 2.x LTS and 3.x now).
> 
> So in case things go wrong, users have a way to downgrade and recover.

LTS versions aren't always packaged, especially if there is no/minimal breakage. The `prometheus` package doesn't have any dependents so this is a good sign that IMO it could just be upgraded to v3 and not have to bifurcate into two packages.

Was there a specific reason you thought we would need both versions packaged?

bryanasdev000 commented:
> > as I have understood, there will be multiple versions being supported at the same time (at least 2, if I understood correctly, the 2.x LTS and 3.x now).
> > So in case things go wrong, users have a way to downgrade and recover.
> 
> LTS versions aren't always packaged, especially if there is no/minimal breakage. The `prometheus` package doesn't have any dependents so this is a good sign that IMO it could just be upgraded to v3 and not have to bifurcate into two packages.
> 
> Was there a specific reason you thought we would need both versions packaged?

Only if the user wants to be in the LTS or if things goes wrong on the update.

Myself will use 3.x

bryanasdev000 commented:
Built 3.0.1 locally and will see how it goes.

Need to fix consoles on 3.0.1 and understand why tests are failing for both, 2.55.1 and 3.0.1 if I rebase to master.

TinfoilSubmarine commented:
[From the prometheus changelog:](https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md)
> [CHANGE] Console: Remove example files for the console feature. Users can continue using the console feature by supplying their own JavaScript and templates. #14807

bryanasdev000 commented:
> [From the prometheus changelog:](https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md)
> 
> > [CHANGE] Console: Remove example files for the console feature. Users can continue using the console feature by supplying their own JavaScript and templates. #14807
> 
> So you can just remove those two lines:
> 
> ```
>         vcopy consoles etc/prometheus
>         vcopy console_libraries etc/prometheus
> ```

I saw this, but it looks like I tested the wrong container image as I swore I saw `/consoles/index.html.example` in 3.0.0.

Anyway, I'll delete these items, thanks.

TinfoilSubmarine commented:
For the test failure, it looks like it's just the one test 'TestQueryLog' that is failing. You could just skip the test:
```
diff --git a/srcpkgs/prometheus/template b/srcpkgs/prometheus/template
index 8d6dd022b46..9faadc82673 100644
--- a/srcpkgs/prometheus/template
+++ b/srcpkgs/prometheus/template
@@ -3,6 +3,7 @@ pkgname=prometheus
 version=3.0.1
 revision=1
 build_style=go
+make_check_args="-skip TestQueryLog"
 go_import_path="github.com/prometheus/prometheus"
 go_package="github.com/prometheus/prometheus/cmd/prometheus github.com/prometheus/prometheus/cmd/promtool"
 go_ldflags="-X github.com/prometheus/common/version.Version=${version}
 ```

bryanasdev000 commented:
> For the test failure, it looks like it's just the one test 'TestQueryLog' that is failing. You could just skip the test:
> 
> ```
> diff --git a/srcpkgs/prometheus/template b/srcpkgs/prometheus/template
> index 8d6dd022b46..9faadc82673 100644
> --- a/srcpkgs/prometheus/template
> +++ b/srcpkgs/prometheus/template
> @@ -3,6 +3,7 @@ pkgname=prometheus
>  version=3.0.1
>  revision=1
>  build_style=go
> +make_check_args="-skip TestQueryLog"
>  go_import_path="github.com/prometheus/prometheus"
>  go_package="github.com/prometheus/prometheus/cmd/prometheus github.com/prometheus/prometheus/cmd/promtool"
>  go_ldflags="-X github.com/prometheus/common/version.Version=${version}
> ```

After a few tries, atleast here (x86_64-glibc) the only failing test is `TestDelayedCompaction`.

TinfoilSubmarine commented:
>  the only failing test is `TestDelayedCompaction`.

Is this issue similar to the failure you were seeing then? https://github.com/prometheus/prometheus/issues/14855

bryanasdev000 commented:
> > the only failing test is `TestDelayedCompaction`.
> 
> Is this issue similar to the failure you were seeing then? [prometheus/prometheus#14855](https://github.com/prometheus/prometheus/issues/14855)

Yes, same output.

bryanasdev000 commented:
> Out of curiosity, I ran this on my own branche's CI: [TinfoilSubmarine#8](https://github.com/TinfoilSubmarine/void-packages/pull/8)
> 
> This is without any skipped tests. The i686 failure is unrelated to failing tests.

Curious, no error on the tests? I don't have a full domain over xbps-src and how it builds the packages, but I have a plenty of free space in the machine that I am using to build the packages (around 80GB at the moment at / and 700GB at /home, where is my copy of void-packages, maybe some tmpfs thing?).

Also, any idea why i686 just seems to stop without logs?

I should drop the skip test on the template?

TinfoilSubmarine commented:
> Curious, no error on the tests? I don't have a full domain over xbps-src and how it builds the packages, but I have a plenty of free space in the machine that I am using to build the packages (around 80GB at the moment at / and 700GB at /home, where is my copy of void-packages, maybe some tmpfs thing?).

The build happens in the chroot located at `masterdir-<arch>` in the `void-packages` directory by default.

> Also, any idea why i686 just seems to stop without logs?

It's just truncated, you can view the whole log if you go up to the top-right and click view raw logs.

> I should drop the skip test on the template?

We may have to do something like this to only run those tests when in CI:
```
if [ "${XBPS_BUILD_ENVIRONMENT}" != "void-packages-ci" ]; then
  make_check_args+="-skip (TestQueryLog|TestDelayedCompaction)"
fi
```

TinfoilSubmarine commented:
Looking at the logs for i686 it seems like this is a 32-bit int/float type size assumption bug.

bryanasdev000 commented:
> > Curious, no error on the tests? I don't have a full domain over xbps-src and how it builds the packages, but I have a plenty of free space in the machine that I am using to build the packages (around 80GB at the moment at / and 700GB at /home, where is my copy of void-packages, maybe some tmpfs thing?).
> 
> The build happens in the chroot located at `masterdir-<arch>` in the `void-packages` directory by default.
> 
> > Also, any idea why i686 just seems to stop without logs?
> 
> It's just truncated, you can view the whole log if you go up to the top-right and click view raw logs.
> 
> > I should drop the skip test on the template?
> 
> We may have to do something like this to only run those tests when in CI:
> 
> ```
> if [ "${XBPS_BUILD_ENVIRONMENT}" != "void-packages-ci" ]; then
>   make_check_args+="-skip (TestQueryLog|TestDelayedCompaction)"
> fi
> ```

Ok, thanks, will add it.

bryanasdev000 commented:
> Looking at the logs for i686 it seems like this is a 32-bit int/float type size assumption bug.

Will you report it upstream? Asking because I really don't know if they support 32-bit arches beside Armv6 (if I remember correctly).

bryanasdev000 commented:
Well, another flaky one here `TestHeadSeriesChunkRace`.

TinfoilSubmarine commented:
> > Looking at the logs for i686 it seems like this is a 32-bit int/float type size assumption bug.
> 
> Will you report it upstream? Asking because I really don't know if they support 32-bit arches beside Armv6 (if I remember correctly).

They do support 386/i686, look at the releases. There's a build for both linux and freebsd. [Apparently](https://github.com/prometheus/prometheus/blob/main/.github/workflows/ci.yml#L36) they only run the tests for `./cmd/prometheus` for 386 though, so that's why these failures haven't been caught.

TinfoilSubmarine commented:
Opened https://github.com/prometheus/prometheus/issues/15547

TinfoilSubmarine commented:
Built and tested on x86_64, working fine.

I took some time to patch all of the 32-bit test bugs here: https://github.com/prometheus/prometheus/pull/15548

And I have applied the patches to my branch here: https://github.com/TinfoilSubmarine/void-packages/pull/8. You could either pull these patches into this branch, or just skip the failing tests on 32-bit for now.

TinfoilSubmarine commented on srcpkgs/prometheus/template:
```diff
@@ -1,8 +1,11 @@
+version=3.0.1
+revision=1
 build_style=go
+if [ "${XBPS_BUILD_ENVIRONMENT}" != "void-packages-ci" ]; then
+  make_check_args+="-skip (TestQueryLog|TestDelayedCompaction)"
+fi
```
This block should be after `make_dirs=...`. It also needs a space at the beginning of the string, that's my bad for not suggesting it correctly:
```
make_check_args+=" -skip (TestQueryLog|TestDelayedCompaction)"
```

bryanasdev000 replied:
No problem, fixed.

bryanasdev000 commented:
> Built and tested on x86_64, working fine.
> 
> I took some time to patch all of the 32-bit test bugs here: [prometheus/prometheus#15548](https://github.com/prometheus/prometheus/pull/15548)
> 
> And I have applied the patches to my branch here: [TinfoilSubmarine#8](https://github.com/TinfoilSubmarine/void-packages/pull/8). You could either pull these patches into this branch, or just skip the failing tests on 32-bit for now.

Sorry for the late response, also merry (late) christmas!

Since you PR is already merged, I prefer to wait for a release from upstream, if that's okay.

I fixed `make_check_args` order, rebased and pushed.

bryanasdev000 commented:
Fixed spaces vs tabs.

bryanasdev000 commented:
Local build ok.

TinfoilSubmarine commented:
3.1.0 is released.

bryanasdev000 commented:
> 3.1.0 is released.

Happy (late) new year!

Built it locally and will test it on my lab, as for tests, seems that a new one (maybe more) is failing (at least here):

```
panic: test timed out after 10m0s
        running tests:
                TestWaitForPendingReadersInTimeRange (3s)
                TestWaitForPendingReadersInTimeRange/mint=500,maxt=1500,shouldWait=true (1s)
```

bryanasdev000 commented:
Well, no errors for tests on CI, but i686 may be borked because of https://github.com/rollup/rollup/issues/5392 (based on CI logs https://github.com/void-linux/void-packages/actions/runs/12626419663/job/35179537404?pr=53119).

TinfoilSubmarine commented:
> Well, no errors for tests on CI, but i686 may be borked because of [rollup/rollup#5392](https://github.com/rollup/rollup/issues/5392) (based on CI logs https://github.com/void-linux/void-packages/actions/runs/12626419663/job/35179537404?pr=53119).

I have a fix for this in `pre_build()` on my branch:
```
pre_build() {
 	# Need to patch node fs to avoid using too many file descriptors
 	(cd web/ui && yarn add --dev -W graceful-fs@4.2.11)
 	echo "require('graceful-fs').gracefulify(require('fs'));" \
 		> "${wrksrc}/web/ui/use-graceful-fs.js"
 	export NODE_OPTIONS="--require ${wrksrc}/web/ui/use-graceful-fs.js"
+
+	# rollup doesn't support native i686, use wasm instead
+	if [ "$XBPS_TARGET_MACHINE" = "i686" ]; then
+		(cd web/ui && npm install --override rollup@npm:@rollup/wasm-node@latest)
+	fi
+
 	CGO_ENABLED=0 GOARCH= make assets assets-compress
}
```

bryanasdev000 commented:
> > Well, no errors for tests on CI, but i686 may be borked because of [rollup/rollup#5392](https://github.com/rollup/rollup/issues/5392) (based on CI logs https://github.com/void-linux/void-packages/actions/runs/12626419663/job/35179537404?pr=53119).
> 
> I have a fix for this in `pre_build()` on my branch:
> ```
> pre_build() {
>  	# Need to patch node fs to avoid using too many file descriptors
>  	(cd web/ui && yarn add --dev -W graceful-fs@4.2.11)
>  	echo "require('graceful-fs').gracefulify(require('fs'));" \
>  		> "${wrksrc}/web/ui/use-graceful-fs.js"
>  	export NODE_OPTIONS="--require ${wrksrc}/web/ui/use-graceful-fs.js"
> +
> +	# rollup doesn't support native i686, use wasm instead
> +	if [ "$XBPS_TARGET_MACHINE" = "i686" ]; then
> +		(cd web/ui && npm install --override rollup@npm:@rollup/wasm-node@latest)
> +	fi
> +
>  	CGO_ENABLED=0 GOARCH= make assets assets-compress
> }
> ```

Done, thanks.

TinfoilSubmarine commented:
This looks good to me. You should squash all of the commits into one with the message matching the title of this PR per [CONTRIBUTING.md#committing-your-changes](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#committing-your-changes) and mark this PR as ready for review.

bryanasdev000 commented:
> This looks good to me. You should squash all of the commits into one with the message matching the title of this PR per [CONTRIBUTING.md#committing-your-changes](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#committing-your-changes) and mark this PR as ready for review.

Done, thanks for the help.

TinfoilSubmarine commented:
3.2.0 is out
